wcfsetup/install/files/lib/action/ModerationQueueAssignUserAction.class.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+namespace wcf\action;
+
+use CuyZ\Valinor\Mapper\Source\Source;
+use CuyZ\Valinor\Mapper\TreeMapper;
+use CuyZ\Valinor\MapperBuilder;
+use Laminas\Diactoros\Response\JsonResponse;
+use Laminas\Diactoros\Response\TextResponse;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use wcf\data\moderation\queue\ModerationQueue;
+use wcf\data\user\User;
+use wcf\http\middleware\Xsrf;
+use wcf\system\exception\IllegalLinkException;
+use wcf\system\exception\InvalidSecurityTokenException;
+use wcf\system\exception\PermissionDeniedException;
+use wcf\system\form\builder\field\dependency\ValueFormFieldDependency;
+use wcf\system\form\builder\field\SingleSelectionFormField;
+use wcf\system\form\builder\field\user\UserFormField;
+use wcf\system\form\builder\FormDocument;
+use wcf\system\moderation\queue\command\AssignUser;
+use wcf\system\WCF;
+use wcf\util\JSON;
+
+/**
+ * Assigns a user to a moderation queue entry.
+ *
+ * @author  Tim Duesterhus
+ * @copyright   2001-2022 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package WoltLabSuite\Core\Action
+ */
+final class ModerationQueueAssignUserAction implements RequestHandlerInterface
+{
+    private const PARAMETERS = <<<'EOT'
+        array {
+            id: positive-int
+        }
+        EOT;
+
+    private TreeMapper $mapper;
+
+    public function __construct()
+    {
+        $this->mapper = (new MapperBuilder())
+            // The query string might contain arbitrary parameters that we want to ignore.
+            ->flexible()
+            ->mapper();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        if (
+            $request->getMethod() === 'POST'
+            && !$request->getAttribute(Xsrf::HAS_VALID_HEADER_ATTRIBUTE, false)
```

Comment:
Maybe add `Xsrf::assertHasXsrfToken($request)` that checks the request for the attribute and throws the exception otherwise?

Reply:
Actually, I would rather opt for a “secure by default” approach.

The actual request should be wrapped in a middleware that checks if it implements `RequestHandlerInterface`. If this true, the XSRF check should be enforced.

An opt-out should be possible through an attribute like `#[DisableXsrfCheck]`.

This not only saves time for most implementations, but also makes it harder to “forget” the check.

Reply:
On second thought, we might need a parameter to opt-out for some request methods only.

```php
#[DisableXsrfCheck]
final class NeverCheckForXsrf implements RequestHandlerInterface
```

```php
#[DisableXsrfCheck('GET')]
final class CheckForXsrfUnlessGET implements RequestHandlerInterface
```

```php
#[DisableXsrfCheck('GET', 'HEAD')]
final class CheckForXsrfUnlessGETorHEAD implements RequestHandlerInterface
```

If present without parameters the attribute will disable the XSRF validation for all methods. A list of HTTP verbs can be provided as a list of string parameters (the attribute consumes `array ...$verbs`), disabling the check for these verbs only. Providing an unknown verb causes an exception.

Reply:
Now resolved with #5124